crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -71,77 +54,53 @@ use crate::{
 
 mod data_ingestion_handler;
 pub mod metrics;
+pub(crate) mod utils;
 
-type CheckpointExecutionBuffer = FuturesOrdered<
-    JoinHandle<(
-        VerifiedCheckpoint,
-        Option<Accumulator>,
-        Option<CheckpointData>,
-        Vec<TransactionDigest>,
-        Vec<RandomnessRound>,
-    )>,
->;
-
-/// The interval to log checkpoint progress, in # of checkpoints processed.
-const CHECKPOINT_PROGRESS_LOG_COUNT_INTERVAL: u64 = 5000;
+use data_ingestion_handler::{load_checkpoint_data, store_checkpoint_locally};
+use metrics::CheckpointExecutorMetrics;
+use utils::*;
 
-#[derive(Debug, Clone, Copy)]
-pub struct CheckpointTimeoutConfig {
-    pub panic_timeout: Option<Duration>,
-    pub warning_timeout: Duration,
-}
+const CHECKPOINT_PROGRESS_LOG_COUNT_INTERVAL: u64 = 5000;
 
-// We use a thread local so that the config can be overridden on a per-test basis. This means
-// that get_scheduling_timeout() can be called multiple times in a multithreaded context, but
-// the function is still very cheap to call so this is okay.
-thread_local! {
-    static SCHEDULING_TIMEOUT: once_cell::sync::OnceCell<CheckpointTimeoutConfig> =
-        const { once_cell::sync::OnceCell::new() };
+#[derive(PartialEq, Eq, Debug)]
+pub enum StopReason {
+    EpochComplete,
+    RunWithRangeCondition,
 }
 
-#[cfg(msim)]
-pub fn init_checkpoint_timeout_config(config: CheckpointTimeoutConfig) {
-    SCHEDULING_TIMEOUT.with(|s| {
-        s.set(config).expect("SchedulingTimeoutConfig already set");
-    });
+pub(crate) struct CheckpointExecutionData {
+    pub checkpoint: VerifiedCheckpoint,
+    pub checkpoint_contents: CheckpointContents,
+    pub tx_digests: Vec<TransactionDigest>,
+    pub fx_digests: Vec<TransactionEffectsDigest>,
+    pub transactions: Vec<VerifiedExecutableTransaction>,
+    pub effects: Vec<TransactionEffects>,
 }
 
-fn get_scheduling_timeout() -> CheckpointTimeoutConfig {
-    fn inner() -> CheckpointTimeoutConfig {
-        let panic_timeout: Option<Duration> = if cfg!(msim) {
-            Some(Duration::from_secs(45))
-        } else {
-            std::env::var("NEW_CHECKPOINT_PANIC_TIMEOUT_MS")
-                .ok()
-                .and_then(|s| s.parse::<u64>().ok())
-                .map(Duration::from_millis)
-        };
+pub(crate) struct CheckpointExecutionState {
+    pub data: CheckpointExecutionData,
 
-        let warning_timeout: Duration = std::env::var("NEW_CHECKPOINT_WARNING_TIMEOUT_MS")
-            .ok()
-            .and_then(|s| s.parse::<u64>().ok())
-            .map(Duration::from_millis)
-            .unwrap_or(Duration::from_secs(5));
+    executed_fx_digests: Vec<Option<TransactionEffectsDigest>>,
```

Comment:
Whats the purpose in having the fx digests repeated here as well as in the CheckpointExecutionData?

I think i can hazard a guess that the CheckpointExecutionData is the data that we've received via statesync and we haven't executed it ourselves yet? Maybe a comment to describe this?

Reply:
There's actually no duplication - CheckpointExecutionData contains the certified state (including effects/effects digests).

This vector is the locally executed digests for each transaction, which are optional because some transactions may not be executed yet. For those that are present, we ensure that they match the certified digests.